### PR TITLE
Release notes generator: RC support and ignore prereleases to generate changelog since previous release

### PR DESCRIPTION
**Pull request by dmitrii-ubskii**
## What is the goal of this PR?

We update the release notes generator tool to be able to handle `-rc1` in a version tag, and to generate the complete change log during a full release since previous full release.

We also only emit a PR description once per PR. Github lists as pulls for a given commit all pull requests involving the commit, which includes e.g. merges from development into master. We now restrict each PR description to only be emitted once, and fallback to the commit message if all PRs have been filtered out.

### CODE REVIEW

File: tool/release/notes/Version.kt [hunk 1]
```diff
@@ -21,65 +21,76 @@
...
+                        "version '$version' does not follow the form 'x.y.z', 'x.y.z-rcN', 'x.y.z-alpha', or 'x.y.z-alpha-N'"
+                    ) 
+                }
             }
         }
     }
 
     override fun compareTo(other: Version): Int {
         val major = major.compareTo(other.major)
-        if (major == 0) {
```
Comment (dmitrii-ubskii) [inline tool/release/notes/Version.kt:60]: Flattened the logic

File: tool/release/notes/Commit.kt [hunk 2]
```diff
@@ -54,10 +54,18 @@ private fun getPrecedingVersion(org: String, repo: String, version: Version, git
     tags.addAll(body.asArray().map { release -> Version.parse(release.asObject().get("tag_name").asString()) })
     tags.sort()
     val currentIdx = tags.indexOf(version)
-    val preceding =
-        if (currentIdx >= 1) tags[currentIdx - 1]
-        else if (currentIdx == 0) null
-        else throw IllegalStateException("Version '$version' not found: currentIdx = '$currentIdx'")
+    val preceding = when {
+        currentIdx < 0 -> throw IllegalStateException("Version '$version' not found: currentIdx = '$currentIdx'")
```
Comment (dmitrii-ubskii) [inline tool/release/notes/Commit.kt:58]: `if {} else if {} else {}` is canonically represented by `when {}` in Kotlin

File: tool/release/notes/Commit.kt [hunk 3]
```diff
@@ -54,10 +54,18 @@ private fun getPrecedingVersion(org: String, repo: String, version: Version, git
...
+        currentIdx == 0 -> null
+        version.isPrerelease() -> tags[currentIdx - 1]
+        else -> {
+            var previousRelease = currentIdx - 1
+            while (previousRelease >= 0 && tags[previousRelease].isPrerelease()) previousRelease--
+
+            if (previousRelease < 0) null
+            else tags[previousRelease]
```
Comment (dmitrii-ubskii) [inline tool/release/notes/Commit.kt:66]: First full release since repo creation

File: tool/release/notes/Version.kt [hunk 4]
```diff
@@ -21,65 +21,76 @@
 
 package com.vaticle.dependencies.tool.release.notes
 
-data class Version(val major: Int, val minor: Int, val patch: Int, val alpha: Int?): Comparable<Version> {
+data class Version(val major: Int, val minor: Int, val patch: Int, val rc: Int?, val alpha: Int?): Comparable<Version> {
```
Comment (dmitrii-ubskii) [inline tool/release/notes/Version.kt:24]: _Technically_ we should have a proper prerelease type for this. Currently it's possible to have both alpha and rc here. That is out of scope for this PR though.

**Review by flyingsilverfin — APPROVED** [review]
